### DISCUSSION TITLE
Question about sendzc notif CQE ordering

### DISCUSSION THREAD

**Discussion by bnbajwa** [axboe/liburing#970] [Q&A] [answered]
Hello,

Consider the following sequence:
1) submit sendzc SQE for buf1
2) wait for 1st sendzc CQE
3) submit sendzc SQE for buf2
4) wait for 1st sendzc CQE

At this point, I expect 2 notif CQEs for each of the two bufs.

My question: is there any ordering guaranteed for the notif CQEs? i.e. is the following sequence guaranteed?:
1) notif CQE for buf1
2) notif CQE for buf2

i.e. could notif CQE for buf2 arrive BEFORE notif CQE for buf1?

Many Thanks

**Comment by axboe** [comment]
CC @isilence

**Comment by redbaron** [comment]
If you have both buffers ready to be sent, then submit them linked  or use single vector write

**Comment by isilence** [comment, marked as answer]
@bnbajwa, there is no ordering guarantee for zc notification CQEs.

> If you have both buffers ready to be sent, then submit them linked or use single vector write

And it wouldn't change it for notifications, because there is no ordering b/w notification and linked requests.
E.g. with `sendzc -> nop` you can likely a completion ordering as this: ```sendzc - nop -  notif```

**Comment by isilence** [comment]
But it's likely to be ordered, so it can still be used as an optimisation if you handle corner cases in the userspace.
